### PR TITLE
[WIP] starts a testing system, name + types for now

### PR DESCRIPTION
**Pull request by BBasile**
This is based on the runnable module i've started to use to help with the auto array bug. The goal of this first test kind is to help enhancing auto deduction from initializers.

### CODE REVIEW

**Comment by BBasile** [comment]
If okay on the principle i'll add more `auto` tests. Maybe also a printer (var name -> type name -> type name ...)

**Comment by BBasile** [comment]
ping @Hackerpilot. This is not harmful so let's say that if Sunday there's no reaction then i'll add more tests and merge. At least this will prevent regressions when the type deduction solver is updated, without the need to use DCD to detect them.

File: src/dsymbol/tests.d [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            assert(i <= results.length,
+                "not enough results");
+            assert(results[i].length > 1,
+                "at least one type must be present in a result row");
+            assert(ss.name == results[i][0],
+                "expected variableName: `%s` but got `%s`".format(results[i][0], ss.name));
+
+            auto t = cast() ss.type;
+            foreach(immutable j; 1..results[i].length)
```
Comment (wilzbach) [inline src/dsymbol/tests.d]: FWIW `zip` at the for loop above gets you the length validation for free (and avoids the need to keep an index)

File: src/dsymbol/tests.d [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        foreach(ss; (*s.symbol)[])
+    {
+        if (ss.type)
+        {
+            assert(i <= results.length,
+                "not enough results");
+            assert(results[i].length > 1,
+                "at least one type must be present in a result row");
+            assert(ss.name == results[i][0],
+                "expected variableName: `%s` but got `%s`".format(results[i][0], ss.name));
```
Comment (wilzbach) [inline src/dsymbol/tests.d]: At Dscanner the pattern is to start with uppercase and end with a dot.

File: src/dsymbol/tests.d [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * Params:
+ *      source = The source code to test.
+ *      results = An array of string array. Each slot represents the variable name
+ *      followed by the type strings.
+ */
+version (unittest)
+void expectSymbolsAndTypes(const string source, const string[][] results)
+{
+    static immutable rName = "dsymbol-test-154251542-56564105-78944416-98523170-02452336.d";
```
Comment (wilzbach) [inline src/dsymbol/tests.d]: Note: Why don't you generate the id dynamically?
Also if the test suite ever gets run on multiple threads, having a hard-code file name might not be a good idea.

Comment (BBasile) [inline src/dsymbol/tests.d]: Yes i could use a GUID. but...

**Comment by BBasile** [comment]
Thx for handling this but i wished to add more unittest !

**Comment by wilzbach** [comment]
> i wished to add more unittest !

No one stops you from adding more :)
As said this is a lot better then no test suite.

**Pull request merged**